pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AstVisitorBase.java

```diff
@@ -26,8 +26,8 @@
      */
     // kept separate from super.visit for clarity
     protected R visitChildren(Node node, P data) {
-        for (Node child : node.children()) {
-            child.acceptVisitor(this, data);
+        for (int i = 0, numChildren = node.getNumChildren(); i < numChildren; i++) {
+            node.getChild(i).acceptVisitor(this, data);
```

Comment:
I think the version as written here is easier to optimize for the JVM. Using `.children()` almost certainly causes an allocation, because that method may return several different implementations of node streams (empty, singleton, children). That makes the implicit `iterator()` call virtual and not inlineable. By contrast, the loops as written here use only final methods of AbstractNode and can be optimized into simple array loops all the time. Since this method is called very frequently I think it's worth keeping it that way.

Alternatively we could make `children` final on AbstractNode too... I'll look into it

Reply:
So I benchmarked this (time taken visiting a big java tree):
```
Benchmark                       (sourceFname)  Mode  Cnt   Score   Error  Units
StreamBench.nonOptChildren  /PLSQLParser.java  avgt   10  19,138 ± 0,125  ms/op
StreamBench.optChildren     /PLSQLParser.java  avgt   10   6,820 ± 0,049  ms/op
StreamBench.optimizedLoop   /PLSQLParser.java  avgt   10   8,847 ± 0,061  ms/op
```
The slowest implementation is the current one with the `children` loop. The explicit loop as written in this PR is more than twice faster (that's "optimizedLoop"). Even faster is a version (here "optChildren") where I made `children` final on AbstractNode, and made its implementation return the same node stream type all the time.